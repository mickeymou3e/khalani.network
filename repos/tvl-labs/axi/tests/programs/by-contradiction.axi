const a: Boolean;

proc by_contradiction(p) {
    !false_elim();
    !mt(p)
}

# Prove law of excluded-middle
let ex := a | ~a;
!dn(!by_contradiction(assume ~ex {
    !by_contradiction(
        assume a {
            !left_either(a, ~a);
            !absurd(ex, ~ex)
        }
    );
    !right_either(a, ~a);
    !absurd(ex, ~ex)
}));

const c: Boolean;
const d: Boolean;

# c or d => not c => d
assume c | d {
    assume ~c {
        !dn(!by_contradiction(assume ~d {
            let if_c_false := assume c {
                !absurd(c, ~c)
            };
            let if_d_false := assume d {
                !absurd(d, ~d)
            };
            !cd(if_c_false, if_d_false)
        }))
    }
}
