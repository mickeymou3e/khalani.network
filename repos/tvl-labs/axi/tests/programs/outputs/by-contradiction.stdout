Result: Nothing
Result: Nothing
Result: Nothing
Result: Proved(or(a, not(a)))
Result: Nothing
Result: Nothing
Result: Proved(if(or(c, d), if(not(c), d)))
